### PR TITLE
[Team 4] Update agent

### PR DESCRIPTION
**Pull request by hussain2603**
# Summary

Please ignore Training Agent since this is only used for our training purposes.
The Final Agent is an evolutionary agent. This is the only agent that will run in the simulations.

## Additional Information

Some final agent configurations are subject to change. The implementation is final. 

## Test Plan

Tested with our own agent and all other agents currently in main.

### CODE REVIEW

File: pkg/agents/team4/finalAgent/agent.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/agent.go:55]:
```suggestion
func initialiseParams(baseAgent *infra.Base) CustomAgentEvoParams {
```

Comment (jzzheng22) [reply pkg/agents/team4/finalAgent/agent.go:55]: ideally fix but whatever

File: pkg/agents/team4/finalAgent/agent.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/agent.go:86]: Better to use `make([]messages.Messages, 0)` for these I think

File: pkg/agents/team4/finalAgent/agent.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/agent.go:173]: Ideally use `make` I think

File: pkg/agents/team4/finalAgent/agent.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/agent.go:250]: Please break this up into functions for readability:

- You can put everything inside the `if a.hasReshuffled()` statement into a function
- The second if statement should be in a function too

File: pkg/agents/team4/finalAgent/agent.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/agent.go:213]: I have no idea what this is comparing - you should probably put the `waitProbability` thing into a separate variable

File: pkg/agents/team4/finalAgent/messages.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/messages.go:97]:
```suggestion
	response := a.CurrPlatFood()-a.params.intendedFoodToTake >= food.FoodType(amount)
```

File: pkg/agents/team4/finalAgent/treaty.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:31]:
```suggestion
```

File: pkg/agents/team4/finalAgent/treaty.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:35]: Prefer moving these comments to above each `case` so it's more readable. Likewise for the rest of the switch statement, but ultimately up to you

File: pkg/agents/team4/finalAgent/treaty.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:99]: The comment is wrong - `msg.Response()` indicates whether or not they signed the treaty?

File: pkg/agents/team4/finalAgent/treaty.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:103]: You could just call `SignTreaty()`

File: pkg/agents/team4/finalAgent/treaty.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:191]: Ideally put this in a separate function and call it inside the case

File: pkg/agents/team4/finalAgent/treaty.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:208]: Ideally put this in a separate function and call it inside the case

File: pkg/agents/team4/finalAgent/treaty.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:189]: Convert to case and combine with GE condition?

File: pkg/agents/team4/finalAgent/treaty.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/treaty.go:206]: Convert to case and combine with GE condition?

File: pkg/agents/team4/finalAgent/trustScoreUpdates.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/trustScoreUpdates.go:87]: Don't need to put this else here

File: pkg/agents/team4/finalAgent/trustScoreUpdates.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/trustScoreUpdates.go:101]: Put a return and get rid of the else

File: pkg/agents/team4/finalAgent/trustScoreUpdates.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/trustScoreUpdates.go:117]: Why do you need this return?

File: pkg/agents/team4/finalAgent/trustScoreUpdates.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/trustScoreUpdates.go:104]: Move the comment above the if for readability

File: pkg/agents/team4/finalAgent/trustScoreUpdates.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/trustScoreUpdates.go:107]: Can you write the condition across two lines for readability?

File: pkg/agents/team4/finalAgent/utils.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/utils.go:39]:
```suggestion
	for _, sentMsg := range a.params.sentMessages { // Iterate through each sent message
		if resMsg.RequestID() == sentMsg.ID() { // matches request to relevant sentID
			return sentMsg
		}
	}
}
```

File: pkg/agents/team4/finalAgent/utils.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/utils.go:45]:
```suggestion
			return
```
More clear to use return over break I think (but check it still works)

File: pkg/agents/team4/finalAgent/utils.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/utils.go:41]: Should probably be a member function?

File: pkg/agents/team4/finalAgent/utils.go
Comment (jzzheng22) [inline pkg/agents/team4/finalAgent/utils.go:30]: Should probably be a member function?

**Review by jzzheng22 — CHANGES_REQUESTED** [review]
ty for submitting pr early bbygrills much appreciated <3

File: pkg/agents/team4/trainingAgent/generateNewBestAgents.py
Comment (woonmoon) [inline pkg/agents/team4/trainingAgent/generateNewBestAgents.py:48]: MINOR: would be more readable in list comprehension `bands = [ (1/number_of_agents_to_use) * (1/(i+1))) for i in range(number_of_agents_to_use) ]`

Comment (woonmoon) [reply pkg/agents/team4/trainingAgent/generateNewBestAgents.py:48]: IMO a lot of the for-loops in this file could be condensed for readability but that's also minor

File: pkg/agents/team4/trainingAgent/getLifeExpectancy.py
Comment (woonmoon) [inline pkg/agents/team4/trainingAgent/getLifeExpectancy.py:19]: MINOR: `logs = [ json.loads(lines[i]) for i in range(len(lines)) ]` - just a suggestion non-necessary.

File: pkg/agents/team4/finalAgent/utils.go
Comment (woonmoon) [inline pkg/agents/team4/finalAgent/utils.go:15]: I think you can remove the `else if`

File: pkg/agents/team4/finalAgent/utils.go
Comment (hussain2603) [reply pkg/agents/team4/finalAgent/utils.go:39]: it needs to have a return at the end for it to compile, that's why we've done it like this

**Review by woonmoon — COMMENTED** [review]
LGTM, clearly following honest treaties - made some minor suggestions but feel free to dismiss. <3

File: pkg/agents/team4/finalAgent/treaty.go
Comment (hussain2603) [reply pkg/agents/team4/finalAgent/treaty.go:189]: they have different conditions in the rest of the statement

File: pkg/agents/team4/finalAgent/utils.go
Comment (hussain2603) [reply pkg/agents/team4/finalAgent/utils.go:41]: I think it didn't make sense for us to have this functions be a member function of the agent

File: pkg/agents/team4/trainingAgent/getLifeExpectancy.py
Comment (hussain2603) [reply pkg/agents/team4/trainingAgent/getLifeExpectancy.py:19]: the training agent was just for us and won't be used in the final simulation :)

File: pkg/agents/team4/finalAgent/utils.go
Comment (jzzheng22) [reply pkg/agents/team4/finalAgent/utils.go:41]: Ah right, I think if you have a `return nil` at the end then it would work.

**Review by jzzheng22 — APPROVED** [review]
I think it's fine, I've left one suggestion but it's fine if you don't want.

Ideally wait for another approval if possible

**Review by SaraFFdez — APPROVED** [review]
LGTM

File: pkg/agents/team4/finalAgent/utils.go
Comment (JaafarRammal) [reply pkg/agents/team4/finalAgent/utils.go:41]: our yeah, retMsg is undefined if the loop doesn't match anything, is that okay with your agent behaviour?

**Review by JaafarRammal — APPROVED** [review]
Only that little nil retMessage comment otherwise GTG!

**Pull request merged**
